### PR TITLE
[SYSDM] Virtual Memory panel : fix of CORE-12486

### PR DESCRIPTION
**Pull request by KyleKatarn31**
## Purpose

_The Virtual Memory Panel is affected by missing element (current pagefile size), some fr-fr translation glitches and the initial & maximum extreme values are not aligned with WinXP without good reasons (specific ROS design). Check on MinSize below 4096MB is missing.
Moreover since XP was designed in times where 1,5 x AvailMemory was far below 4096 MB, it may recommend a value for pagefile higher than the maximal allowed size. I suggest to cap Recommended value to limit of Maximal value (4096 MB max)

JIRA issue: [CORE-12486](https://jira.reactos.org/browse/CORE-12486)

## Proposed changes

_ Re-enabled computation of current pagefile size, replacing arbitrary page file count (26 static) by the dynamic one used for all other elements
_ Fixed fr-fr translation in order to match Windows XP fr-fr wording and fix typo
_ Minimal value for initial size should be 2 MB, not 16 MB and below 4096 MB
_ Maximal value for maximal size should be 4096 MB, not 4095 MB
_ Capping Recommended value to the limit value for Maximal value

Post change screenshot : 
image

### CODE REVIEW

File: dll/cpl/sysdm/virtmem.c [hunk 1]
```diff
@@ -442,8 +442,9 @@ OnSet(PVIRTMEM pVirtMem)
             }
 
             /* Check the valid range of the minimum size */
-            if (MinSize < 16 ||
-                MinSize > pVirtMem->Pagefile[DriveIndex].FreeSize)
+            if (MinSize < 2 ||
+                MinSize > pVirtMem->Pagefile[DriveIndex].FreeSize ||
+				MinSize > 4096)
```
Comment (sdever) [inline dll/cpl/sysdm/virtmem.c:447]: Use spaces instead of tabs

File: dll/cpl/sysdm/virtmem.c [hunk 2]
```diff
@@ -581,15 +582,13 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
         }
 
         /* Set current pagefile size */
-#if 0
         PageFileSizeMb = 0;
-        for (i = 0; i < 26; i++)
+        for (i = 0; i < pVirtMem->Count; i++)
```
Comment (sdever) [inline dll/cpl/sysdm/virtmem.c:586]: i is not defined according to buildlogs, thus the failed builds.

File: dll/cpl/sysdm/virtmem.c [hunk 3]
```diff
@@ -581,15 +582,13 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
         {
             PageFileSizeMb += pVirtMem->Pagefile[i].InitialSize;
```
Comment (sdever) [inline dll/cpl/sysdm/virtmem.c:588]: InitialSize attribute doesn't seem to exist, see https://travis-ci.org/github/reactos/reactos/builds/676968535#L643. Also checking your changes don't break the build (and work as excepted) is a good idea in general.

Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:588]: I may finally remove this part of the evolution and open a later PR on the "Current pagefile size" part of this PR. Yes, i'll work on this tomorrow.

**Comment by learn-more** [comment]
Please stop spamming untested PR's

File: dll/cpl/sysdm/virtmem.c (hunk 1 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c]: Done

**Comment by KyleKatarn31** [comment]
I've just found some display glitches during my last test... I have to update .rc file

**Comment by KyleKatarn31** [comment]
Everything works EXCEPT the "currently used" section.
2 questions :
- Why is DPRINT1 only logging the 1st letter of the file path and not the full path ?
- Why is GetFileAttributesExW failing ?

Any help is welcome !

**Comment by katahiromz** [comment]
>Why is DPRINT1 only logging the 1st letter of the file path and not the full path?

Due to Unicode string.

**Comment by katahiromz** [comment]
>Why is GetFileAttributesExW failing ?

The 3rd parameter of `GetFileAttributesEx` might be a `WIN32_FILE_ATTRIBUTE_DATA` structure data.
Not a `WIN32_FIND_DATA`.

**Comment by KyleKatarn31** [comment]
> 
> 
> > Why is GetFileAttributesExW failing ?
> 
> The 3rd parameter of `GetFileAttributesEx` might be a `WIN32_FILE_ATTRIBUTE_DATA` structure data.
> Not a `WIN32_FIND_DATA`.

Thank you. With your feedback and last commit the DPRINT1 work OK but the GetFileAttributesExW call is still KO :-(
Any clue ??

File: dll/cpl/sysdm/virtmem.c [hunk 4]
```diff
@@ -502,9 +506,12 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
     TCHAR szBuffer[64];
     MEMORYSTATUSEX MemoryStatus;
     ULARGE_INTEGER FreeDiskSpace;
-    UINT /*i,*/ FreeMemMb /*, PageFileSizeMb*/;
+    UINT i, FreeMemMb, RecoMemMb, PageFileSizeMb;
     INT Index;
-
+    TCHAR szText[MAX_PATH];    
+    WIN32_FIND_DATAW fdata = {0};
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:512]: Fix me!

**Comment by KyleKatarn31** [comment]
Ready for review

File: dll/cpl/sysdm/virtmem.c (hunk 4 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:512]: It's WIN32_FIND_DATAW.

Comment (katahiromz) [reply dll/cpl/sysdm/virtmem.c:512]: Use WIN32_FILE_ATTRIBUTE_DATA instead.

Comment (katahiromz) [reply dll/cpl/sysdm/virtmem.c:512]: OK.

File: dll/cpl/sysdm/virtmem.c [hunk 5]
```diff
@@ -569,27 +576,45 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
+        SetDlgItemText(hwndDlg, IDC_MINIMUM, _T("2 MB"));
 
         /* Set recommended pagefile size */
         MemoryStatus.dwLength = sizeof(MEMORYSTATUSEX);
         if (GlobalMemoryStatusEx(&MemoryStatus))
         {
             FreeMemMb = (UINT)(MemoryStatus.ullTotalPhys / (1024 * 1024));
-            _stprintf(szBuffer, _T("%u MB"), FreeMemMb + (FreeMemMb / 2));
+            RecoMemMb =  FreeMemMb + (FreeMemMb / 2);
+            if(RecoMemMb>4096) RecoMemMb = 4096;
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:587]:
```suggestion
            if (RecoMemMb > 4096)
                RecoMemMb = 4096;
```

File: dll/cpl/sysdm/virtmem.c [hunk 6]
```diff
@@ -569,27 +576,45 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
-        SetDlgItemText(hwndDlg, IDC_MINIMUM, _T("16 MB"));
+        SetDlgItemText(hwndDlg, IDC_MINIMUM, _T("2 MB"));
 
         /* Set recommended pagefile size */
         MemoryStatus.dwLength = sizeof(MEMORYSTATUSEX);
         if (GlobalMemoryStatusEx(&MemoryStatus))
         {
             FreeMemMb = (UINT)(MemoryStatus.ullTotalPhys / (1024 * 1024));
-            _stprintf(szBuffer, _T("%u MB"), FreeMemMb + (FreeMemMb / 2));
+            RecoMemMb =  FreeMemMb + (FreeMemMb / 2);
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:586]:
```suggestion
            RecoMemMb = FreeMemMb + (FreeMemMb / 2);
```

File: dll/cpl/sysdm/virtmem.c [hunk 7]
```diff
@@ -569,27 +576,45 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
+            if(RecoMemMb>4096) RecoMemMb = 4096;
+            _stprintf(szBuffer, _T("%u MB"),RecoMemMb);
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:588]:
```suggestion
            _stprintf(szBuffer, _T("%u MB"), RecoMemMb);
```

Comment (katahiromz) [reply dll/cpl/sysdm/virtmem.c:588]: It's better to use `shlwapi!StrFormatKBSizeW` to format the byte sizes.

File: dll/cpl/sysdm/virtmem.c [hunk 8]
```diff
@@ -569,27 +576,45 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
 
         /* Set current pagefile size */
-#if 0
         PageFileSizeMb = 0;
-        for (i = 0; i < 26; i++)
+        		
+        for (i = 0; i < pVirtMem->Count; i++)
         {
-            PageFileSizeMb += pVirtMem->Pagefile[i].InitialSize;
+		    _stprintf(szText,
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:597]: Unindent.

File: dll/cpl/sysdm/virtmem.c [hunk 9]
```diff
@@ -569,27 +576,45 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
+		    {
+		        ULONGLONG Size = fdata.nFileSizeLow;
+			    Size |= (((__int64)fdata.nFileSizeHigh) << 32);
+			    Size = Size / (1024*1024);
+			    PageFileSizeMb +=Size;
+			    FindClose(hFind);
+		    }
         }
+		
         _stprintf(szBuffer, _T("%u MB"), PageFileSizeMb);
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:616]: Better to use `shlwapi!StrFormatKBSizeW`.

File: dll/cpl/sysdm/virtmem.c [hunk 10]
```diff
@@ -569,27 +576,45 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
+        		
+        for (i = 0; i < pVirtMem->Count; i++)
         {
-            PageFileSizeMb += pVirtMem->Pagefile[i].InitialSize;
+		    _stprintf(szText,
+                          _T("%c:\\pagefile.sys"),
+                          pVirtMem->Pagefile[i].szDrive[0]);
+		
+		    hFind = FindFirstFileW(szText, &fdata);
+		    if (hFind == INVALID_HANDLE_VALUE)
```
Comment (learn-more) [inline dll/cpl/sysdm/virtmem.c:602]: again, mixed tabs and spaces...

File: dll/cpl/sysdm/virtmem.c (hunk 8 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:598]: OK

File: dll/cpl/sysdm/virtmem.c (hunk 10 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c]: fixed... I have to change my notepad++ settings....

File: dll/cpl/sysdm/virtmem.c (hunk 8 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:598]: Done

File: dll/cpl/sysdm/virtmem.c (hunk 10 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c]: fixed

File: dll/cpl/sysdm/virtmem.c (hunk 9 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:617]: Not the right function : I want MB in all cases (not KB nor GB)

File: dll/cpl/sysdm/virtmem.c [hunk 11]
```diff
@@ -569,27 +576,46 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
 
         /* Set current pagefile size */
-#if 0
         PageFileSizeMb = 0;
-        for (i = 0; i < 26; i++)
+        		
+        for (i = 0; i < pVirtMem->Count; i++)
         {
-            PageFileSizeMb += pVirtMem->Pagefile[i].InitialSize;
+		    _stprintf(szText,
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:598]: This line contains tabs.

File: dll/cpl/sysdm/virtmem.c [hunk 12]
```diff
@@ -569,27 +576,46 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
+                      pVirtMem->Pagefile[i].szDrive[0]);
+		
+            hFind = FindFirstFileW(szText, &fdata);
+            if (hFind == INVALID_HANDLE_VALUE) 
+            {
+                DPRINT1("Unable to read PageFile size : %ls due to error %d\n", szText,GetLastError());				
+            }
+            else
+            {
+                ULONGLONG Size = fdata.nFileSizeLow;
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:609]: Use `ULARGE_INTEGER` for 64-bit integer handling.

File: dll/cpl/sysdm/virtmem.c [hunk 13]
```diff
@@ -442,8 +445,9 @@ OnSet(PVIRTMEM pVirtMem)
             }
 
             /* Check the valid range of the minimum size */
-            if (MinSize < 16 ||
-                MinSize > pVirtMem->Pagefile[DriveIndex].FreeSize)
+            if (MinSize < 2 ||
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:448]: Please define a macro (such as `MINSIZE_MIN`) for this magic number.

File: dll/cpl/sysdm/virtmem.c [hunk 14]
```diff
@@ -442,8 +445,9 @@ OnSet(PVIRTMEM pVirtMem)
...
+                MinSize > pVirtMem->Pagefile[DriveIndex].FreeSize ||
+                MinSize > 4096)
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:450]: Please define a macro (such as `MINSIZE_MAX`) for this magic number.

File: dll/cpl/sysdm/virtmem.c [hunk 15]
```diff
@@ -456,7 +460,7 @@ OnSet(PVIRTMEM pVirtMem)
             /* Check the valid range of the maximum size */
             if (MaxSize < MinSize ||
                 MaxSize > pVirtMem->Pagefile[DriveIndex].FreeSize ||
-                MaxSize > 4095)
+                MaxSize > 4096)
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:463]: Please define a macro (such as `MAXSIZE`) for this magic number.

File: dll/cpl/sysdm/virtmem.c (hunk 11 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:598]: i'll fix.

File: dll/cpl/sysdm/virtmem.c (hunk 12 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:609]: OK

File: dll/cpl/sysdm/virtmem.c (hunk 14 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:450]: No benefit (used once)

File: dll/cpl/sysdm/virtmem.c (hunk 15 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:463]: No benefit (used once)

File: dll/cpl/sysdm/virtmem.c (hunk 13 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:448]: No benefit (used once)

File: dll/cpl/sysdm/virtmem.c [hunk 16]
```diff
@@ -569,27 +576,46 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
+                      pVirtMem->Pagefile[i].szDrive[0]);
+        
+            hFind = FindFirstFileW(szText, &fdata);
+            if (hFind == INVALID_HANDLE_VALUE) 
+            {
+                DPRINT1("Unable to read PageFile size : %ls due to error %d\n", szText,GetLastError());                
+            }
+            else
+            {
+                ULARGE_INTEGER Size = fdata.nFileSizeLow;
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:609]: Compilation error

File: dll/cpl/sysdm/virtmem.c (hunk 12 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c]: Cause compilation error :
../dll/cpl/sysdm/virtmem.c:609:17: error: invalid initializer
../dll/cpl/sysdm/virtmem.c:610:22: error: invalid operands to binary | (have 'ULARGE_INTEGER' and 'long long int')
../dll/cpl/sysdm/virtmem.c:611:29: error: invalid operands to binary / (have 'ULARGE_INTEGER' and 'int')
../dll/cpl/sysdm/virtmem.c:612:32: error: invalid operands to binary + (have 'UINT' and 'ULARGE_INTEGER')

File: dll/cpl/sysdm/virtmem.c (hunk 16 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c]: Yes, back to ULONGLONG

Comment (katahiromz) [reply dll/cpl/sysdm/virtmem.c]: Can't you understand `ULARGE_INTEGER`?

Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c]: Yes I do. Is it preferable to the current implementation ?

Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c]: Change towards ULARGE_INTEGER in progress...

**Comment by katahiromz** [comment]
Notice:
- Please avoid mixing translation and real code in one PR.
- Hyphen and Underline are different characters.
- You are noisy.

**Comment by KyleKatarn31** [comment]
> 
> 
> Notice:
> 
>     * Please avoid mixing translation and real code in one PR.
> 
>     * Hyphen and Underline are different characters.
> 
>     * You are noisy.

Thank you for approval
- OK for non-mixing => I'll split in the future
- I know. Where did I mix ?
- I know, due to a PR started before I got RosBE installed. Will be more straightforward next time.

File: dll/cpl/sysdm/virtmem.c [hunk 17]
```diff
@@ -569,27 +577,45 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
         }
 
         /* Set minimum pagefile size */
-        SetDlgItemText(hwndDlg, IDC_MINIMUM, _T("16 MB"));
+        SetDlgItemText(hwndDlg, IDC_MINIMUM, _T("2 MB"));
```
Comment (sdever) [inline dll/cpl/sysdm/virtmem.c:580]: Is there a reason to change this ?

File: dll/cpl/sysdm/virtmem.c [hunk 18]
```diff
@@ -569,27 +577,45 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
 
         /* Set recommended pagefile size */
         MemoryStatus.dwLength = sizeof(MEMORYSTATUSEX);
         if (GlobalMemoryStatusEx(&MemoryStatus))
         {
             FreeMemMb = (UINT)(MemoryStatus.ullTotalPhys / (1024 * 1024));
-            _stprintf(szBuffer, _T("%u MB"), FreeMemMb + (FreeMemMb / 2));
+            RecoMemMb = FreeMemMb + (FreeMemMb / 2);
+            if (RecoMemMb > 4096)
+                RecoMemMb = 4096;
```
Comment (sdever) [inline dll/cpl/sysdm/virtmem.c:589]: Where does this come from ? Do you have tests for this ? Also, the suggestion of @katahiromz is also reused here :D

File: dll/cpl/sysdm/virtmem.c [hunk 19]
```diff
@@ -569,27 +577,45 @@ OnSelChange(HWND hwndDlg, PVIRTMEM pVirtMem)
...
-        SetDlgItemText(hwndDlg, IDC_MINIMUM, _T("16 MB"));
+        SetDlgItemText(hwndDlg, IDC_MINIMUM, _T("2 MB"));
 
         /* Set recommended pagefile size */
         MemoryStatus.dwLength = sizeof(MEMORYSTATUSEX);
         if (GlobalMemoryStatusEx(&MemoryStatus))
         {
             FreeMemMb = (UINT)(MemoryStatus.ullTotalPhys / (1024 * 1024));
-            _stprintf(szBuffer, _T("%u MB"), FreeMemMb + (FreeMemMb / 2));
+            RecoMemMb = FreeMemMb + (FreeMemMb / 2);
```
Comment (katahiromz) [inline dll/cpl/sysdm/virtmem.c:587]: Can you elaborate this line?

Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:587]: This was present before the change. Recommended VM size was 150% x Free memory.
I kept this unchanged. Looks similar to Windows settings

File: dll/cpl/sysdm/virtmem.c (hunk 17 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:580]: Yes, in Windows minimal value is 2 MB. No reason to request more with ROS.

File: dll/cpl/sysdm/virtmem.c (hunk 18 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:589]: This was present before the change. Recommended VM size was 150% x Free memory.
I kept this unchanged. Looks similar to Windows settings

File: dll/cpl/sysdm/virtmem.c (hunk 17 above)
Comment (sdever) [reply dll/cpl/sysdm/virtmem.c:580]: Maybe our borked memory manager could require more (but I'm not even sure that pagefile actually works in ReactOS). Maybe you can try with this value on ReactOS to see if it doesn't break anything.

File: dll/cpl/sysdm/virtmem.c (hunk 19 above)
Comment (katahiromz) [reply dll/cpl/sysdm/virtmem.c:587]: Insert a comment: /* The recommended VM size is 150% of free memory. */

Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:587]: OK. Commit is on its way

File: dll/cpl/sysdm/virtmem.c (hunk 17 above)
Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:580]: Of course I did :)

Comment (learn-more) [inline dll/cpl/sysdm/virtmem.c:580]: In your description you list 'hardcoded strings that need to be localized ("MB") but this is still not localizable.

Comment (KyleKatarn31) [reply dll/cpl/sysdm/virtmem.c:580]: Will be fixed in a separate ticket PR later https://jira.reactos.org/browse/CORE-16915
Description updated (good point)

File: dll/cpl/sysdm/lang/fr-FR.rc [hunk 20]
```diff
@@ -330,11 +330,11 @@ BEGIN
     IDS_USERPROFILE_ACCOUNT_UNKNOWN "Account Unknown"
     IDS_USERPROFILE_ACCOUNT_DELETED "Account Deleted"
     IDS_USERPROFILE_TYPE_TEXT "When %s logs onto this computer, should the operating system use the roaming profile or just the locally cached copy of the roaming profile."
-    IDS_MESSAGEBOXTITLE "System control panel applet"
+    IDS_MESSAGEBOXTITLE "Application Système du Panneau de configuration"
     IDS_WARNINITIALSIZE "Entrez une valeur numérique pour la taille initiale du fichier d'échange."
     IDS_WARNMAXIMUMSIZE "Entrez une valeur numérique pour la taille maximale du fichier d'échange."
-    IDS_WARNINITIALRANGE "La taille initiale du fichier d'échange ne peut pas être inférieure à 16 Mo and ne doit pas dépasser l'espace libre total du disque."
-    IDS_WARNMAXIMUMRANGE "La taille maximale du fichier d'échange ne peut pas être inférieure à sa taille initiale, et ne peut pas être supérieure à 4095 Mo et ne peut pas dépasser l'espace libre total du disque."
+    IDS_WARNINITIALRANGE "La taille initiale du fichier d'échange doit être comprise entre 2 Mo et 4096 Mo, et ne peut pas excéder la quantité d'espace libre sur le lecteur que vous avez sélectionné."
```
Comment (binarymaster) [inline dll/cpl/sysdm/lang/fr-FR.rc:336]: I wonder how this change slipped into master without being synced with other translations, even with default english one. https://github.com/reactos/reactos/blob/756152936b1097a3ee228b1d3c5e364a7c1aeec4/dll/cpl/sysdm/lang/en-US.rc#L336

Comment (KyleKatarn31) [reply dll/cpl/sysdm/lang/fr-FR.rc:336]: I am preparing a complementary PR

Comment (KyleKatarn31) [reply dll/cpl/sysdm/lang/fr-FR.rc:336]: https://github.com/reactos/reactos/pull/2706 : To be approved and merged in fast lane if you consider it is a regression to be fixed.

**Pull request merged**